code/game/machinery/autolathe.dm

```diff
@@ -78,63 +78,118 @@
-				if(ispath(D.build_path, /obj/item/stack))
-					design["max_multiplier"] = min(D.maxstack, D.materials[MAT_METAL] ? round(materials.amount(MAT_METAL) / D.materials[MAT_METAL]) : INFINITY, D.materials[MAT_GLASS] ? round(materials.amount(MAT_GLASS) / D.materials[MAT_GLASS]) : INFINITY)
-				else
-					design["max_multiplier"] = null
-				design["materials"] = design_cost_data(D)
-
-	data = queue_data(data)
+	data["busyname"] =  FALSE
+	data["busyamt"] = 1
+	if(length(being_built) > 0)
+		var/datum/design/D = being_built[1]
+		data["busyname"] =  istype(D) && D.name ? D.name : FALSE
+		data["busyamt"] = length(being_built) > 1 ? being_built[2] : 1
+	data["showhacked"] = hacked ? TRUE : FALSE
+	data["buildQueue"] = queue
+	data["buildQueueLen"] = queue.len
 	return data
 
+/obj/machinery/autolathe/tgui_act(action, list/params, datum/tgui/ui, datum/tgui_state/state)
+	if(..())
+		return FALSE
+
+	add_fingerprint(usr)
+
+	. = TRUE
+	switch(action)
+		if("clear_queue")
+			queue = list()
+		if("remove_from_queue")
+			var/index = text2num(params["remove_from_queue"])
+			if(isnum(index) && ISINRANGE(index, 1, queue.len))
+				remove_from_queue(index)
+				to_chat(usr, "<span class='notice'>Removed item from queue.</span>")
+		if("make")
+			BuildTurf = loc
+			var/datum/design/design_last_ordered
+			design_last_ordered = locateUID(params["make"])
+			if(!istype(design_last_ordered))
+				to_chat(usr, "<span class='warning'>Invalid design</span>")
+				return
+			if(!(design_last_ordered.build_type & AUTOLATHE))
+				to_chat(usr, "<span class='warning'>Invalid design (not buildable in autolathe, report this error.)</span>")
+				return
+			var/datum/component/material_container/materials = GetComponent(/datum/component/material_container)
+			if(design_last_ordered.materials["$metal"] > materials.amount(MAT_METAL))
+				to_chat(usr, "<span class='warning'>Invalid design (not enough metal)</span>")
+				return
+			if(design_last_ordered.materials["$glass"] > materials.amount(MAT_GLASS))
+				to_chat(usr, "<span class='warning'>Invalid design (not enough glass)</span>")
+				return
+			if(!hacked && ("hacked" in design_last_ordered.category))
+				to_chat(usr, "<span class='warning'>Invalid design (lathe requires hacking)</span>")
+				return
+			//multiplier checks : only stacks can have one and its value is 1, 10 ,25 or max_multiplier
+			var/multiplier = text2num(params["multiplier"])
+			var/max_multiplier = min(design_last_ordered.maxstack, design_last_ordered.materials[MAT_METAL] ?round(materials.amount(MAT_METAL)/design_last_ordered.materials[MAT_METAL]):INFINITY,design_last_ordered.materials[MAT_GLASS]?round(materials.amount(MAT_GLASS)/design_last_ordered.materials[MAT_GLASS]):INFINITY)
+			var/is_stack = ispath(design_last_ordered.build_path, /obj/item/stack)
+
+			if(!is_stack && (multiplier > 1))
+				return
+			if(!(multiplier in list(1, 10, 25, max_multiplier))) //"enough materials ?" is checked in the build proc
+				return
```

Comment:
I feel like a `message_admins` here would be good to track potential attempts at href exploits

Reply:
Added